File: scripts/lib/repo-context.mjs
/**
 * Shared helpers for building bounded, full-content repo context. Both scripts
 * run after a full `actions/checkout`, so — unlike a diff-only review —
 * they can hand the model the complete text of the files that matter
 * instead of just paths or diff hunks.
 */
import fs from "node:fs";
import path from "node:path";

const DEFAULT_EXCLUDE_DIRS = new Set([
  ".git",
  "node_modules",
  "vendor",
  "dist",
  ".kiln",
]);

/** Walk `root` and return an indented tree string, skipping noisy dirs. */
export function buildRepoTree(root, { excludeDirs = DEFAULT_EXCLUDE_DIRS, maxEntries = 1500 } = {}) {
  const lines = [];
  let count = 0;

  function walk(dir, depth) {
    if (count >= maxEntries) return;
    let entries;
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    entries.sort((a, b) => a.name.localeCompare(b.name));
    for (const entry of entries) {
      if (count >= maxEntries) return;
      if (entry.name === ".git") continue;
      if (entry.isDirectory() && excludeDirs.has(entry.name)) continue;
      lines.push(`${"  ".repeat(depth)}${entry.name}${entry.isDirectory() ? "/" : ""}`);
      count++;
      if (entry.isDirectory()) walk(path.join(dir, entry.name), depth + 1);
    }
  }

  walk(root, 0);
  if (count >= maxEntries) lines.push(`... (truncated at ${maxEntries} entries)`);
  return lines.join("\n");
}

/** Read a file's content, capped at maxBytes, returning null if missing/unreadable. */
export function readFileSafe(filePath, maxBytes = 20_000) {
  try {
    const buf = fs.readFileSync(filePath);
    if (buf.length <= maxBytes) return buf.toString("utf8");
    return `${buf.subarray(0, maxBytes).toString("utf8")}\n... (truncated, ${buf.length} bytes total)`;
  } catch {
    return null;
  }
}

/**
 * Assemble labeled full-text sections for `relativePaths` under `root`,
 * stopping once `maxTotalBytes` is reached rather than growing unbounded.
 */
export function collectFileContents(root, relativePaths, { maxTotalBytes = 300_000, maxFileBytes = 40_000 } = {}) {
  const sections = [];
  let used = 0;
  for (const rel of relativePaths) {
    if (used >= maxTotalBytes) {
      sections.push(`(skipped remaining files — repo context byte limit of ${maxTotalBytes} reached)`);
      break;
    }
    const content = readFileSafe(path.join(root, rel), maxFileBytes);
    if (content === null) continue;
    const section = `--- ${rel} ---\n${content}`;
    used += section.length;
    sections.push(section);
  }
  return sections.join("\n\n");
}

File: tools/triage-issue/repocontext.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

var defaultExcludeDirs = map[string]bool{
	".git":   true,
	"vendor": true,
	".kiln":  true,
}

const defaultMaxTreeEntries = 1500

// buildRepoTree walks root and returns an indented tree string, skipping
// noisy directories
func buildRepoTree(root string) string {
	var lines []string
	count := 0

	var walk func(dir string, depth int)
	walk = func(dir string, depth int) {
		if count >= defaultMaxTreeEntries {
			return
		}
		entries, err := os.ReadDir(dir)
		if err != nil {
			return
		}
		sort.Slice(entries, func(i, j int) bool {
			return entries[i].Name() < entries[j].Name()
		})
		for _, entry := range entries {
			if count >= defaultMaxTreeEntries {
				return
			}
			if entry.Name() == ".git" {
				continue
			}
			if entry.IsDir() && defaultExcludeDirs[entry.Name()] {
				continue
			}
			suffix := ""
			if entry.IsDir() {
				suffix = "/"
			}
			lines = append(lines, strings.Repeat("  ", depth)+entry.Name()+suffix)
			count++
			if entry.IsDir() {
				walk(filepath.Join(dir, entry.Name()), depth+1)
			}
		}
	}

	walk(root, 0)
	if count >= defaultMaxTreeEntries {
		lines = append(lines, fmt.Sprintf("... (truncated at %d entries)", defaultMaxTreeEntries))
	}
	return strings.Join(lines, "\n")
}

// readFileSafe reads a file's content capped at maxBytes, returning
// (content, true) on success or ("", false) if the file is missing or
// unreadable -- a port of readFileSafe(), using an (string, bool) result
// instead of JS's null-sentinel since Go has no natural "nullable string".
func readFileSafe(filePath string, maxBytes int) (string, bool) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return "", false
	}
	if len(data) <= maxBytes {
		return string(data), true
	}
	return fmt.Sprintf("%s\n... (truncated, %d bytes total)", string(data[:maxBytes]), len(data)), true
}

const defaultMaxFileBytes = 40_000

// collectFileContents assembles labeled full-text sections for
// relativePaths under root, stopping once maxTotalBytes is reached --
// a port of collectFileContents().
func collectFileContents(root string, relativePaths []string, maxTotalBytes int) string {
	var sections []string
	used := 0
	for _, rel := range relativePaths {
		if used >= maxTotalBytes {
			sections = append(sections, fmt.Sprintf("(skipped remaining files — repo context byte limit of %d reached)", maxTotalBytes))
			break
		}
		content, ok := readFileSafe(filepath.Join(root, rel), defaultMaxFileBytes)
		if !ok {
			continue
		}
		section := fmt.Sprintf("--- %s ---\n%s", rel, content)
		used += len(section)
		sections = append(sections, section)
	}
	return strings.Join(sections, "\n\n")
}
